### ISSUE TITLE
Filesystem differences when round-trip rebuilding 999

### ISSUE THREAD

**Issue by PhoenixBound** [AetiasHax/ds-rom#6]
Since this library has "The built ROMs should exactly match the originals" as a goal, I tested running the CLI tool's `extract` and `build` commands on the 2009 game "999: Nine Hours, Nine Persons, Nine Doors" (tested with both BSKJ and BSKE). The ROM did extract and build, and luckily the code was all compressed the same way, but there were a few differences in the filesystem (and ROM header, as a result):

* There were some diffs in the file name table, seemingly caused by path_order.txt being simplified incorrectly. ~~(While investigating the first diff in the table, I noticed the filenames in the bg/a01/outline folder were not originally stored in lexicographic order, in either of Shift-JIS or Unicode, but path_order.txt only included the folder's name and none of the files inside.) I'm trying to put together a minimal example with that reproduces this right now.~~
    * EDIT: After constructing this [minimal project example](https://github.com/user-attachments/files/17158675/test-999-like.zip) (should have nothing copyrighted by Nintendo or Chunsoft, except file names), I realized that the intent seems to be for path\_order.txt to affect the order of files at the end of the *ROM*, not in the *file name table*. So, uh, I guess the cause is just the tool assuming that the FNT has to be sorted... which might be reasonable for most ROMs. It seems to be in order for the rest of the games I've tested so far.
* There was also a difference in the final reported size of the ROM, for some reason... the file allocation table is reported to be the same size, but the file data extends further in the ROM. This seems like it might be related to files being in different order, but I haven't really checked the details yet.

I know there's a few weird things in 999 that will probably merit a fork of this library (mainly, that all Shift-JIS characters with a second byte of `5C` got replaced with different characters in file names -- I'd like to have the original file names with no substituted characters and no external tools for renaming), but I thought these issues *sounded* like generally useful things that might not be specific to 999.

**Comment by AetiasHax** [comment]
I think I've got it! FNT ordering is case-insensitive which can be accomplished by checking if a character is within the range A..Z (or 0x41..0x5a) and then adding 0x20 to conform it to the range a..z. But it looks like they are doing the same thing with each byte in multibyte Shift-JIS characters.

For example, the character コ is encoded as 0x83 0x52. The 0x83 is not in 0x41..0x5a so it is untouched. The 0x52 is however in this range, so 0x20 is added to it and it becomes 0x72, which effectively turns コ into ビ. And you can see that there is a file called *ビン.b3d* that is sandwiched between a few other files starting with コ, since they were treated as equal characters.

> There was also a difference in the final reported size of the ROM, for some reason... the file allocation table is reported to be the same size, but the file data extends further in the ROM. This seems like it might be related to files being in different order, but I haven't really checked the details yet.

This can happen if the last appended file is wrong, as the `rom_size_ds` field does not account for padding. But `ds-rom` still got the last file right so I'll have to investigate this further.

> all Shift-JIS characters with a second byte of 5C got replaced with different characters in file names

I couldn't find any file names with 5C in the FNT, am I missing something?

**Comment by PhoenixBound** [comment]
Oh, that's really interesting! And it makes a lot of sense now that you explained how the sorting confusion happened. Thank you for looking into it.

> > all Shift-JIS characters with a second byte of 5C got replaced with different characters in file names
> 
> I couldn't find any file names with 5C in the FNT, am I missing something?

Oh, uh, I meant like internally. 999's scripts will say something like "display the `予備トイレットペーパー` model now," but then a wrapper function, seemingly written by Chunsoft, sees the 予 character (`97 5C`) and internally replaces it with a different character (`β` in this case -- the mapping isn't done mathematically AFAIK, there's just a bunch of hardcoded character pairings) and lowercases everything before calling out to Nintendo's filesystem code. This also means that the FNT contains that modified name (`β備トイレットペーパー`) instead of the file name written in the script.

In the future, I'd like to include that renaming step in the ROM dump/rebuild tool, to avoid copying or renaming files in a separate step, and so that dumping the files would produce the same filenames that the scripts use (excluding case distinction issues). But I'm not aware of any game besides 999 that would benefit from having a file name modification step before opening or saving the file... so I was just planning to hardcode that into the tool/library myself when the time comes.

**Comment by AetiasHax** [comment]
Interesting... I do wonder if it has to do with 5C being a backslash `\` in ASCII, seeing as backslashes are not allowed in file names and not a single file in the FNT has 5C in its name. But I also don't see any bitwise connection between 予 (97 5C) and β (83 C0), so maybe the developers were unable to use 予 and were forced to change it to an arbitrary character.

**Comment by PhoenixBound** [comment]
There are other characters, yeah. That was just the first mapping I was able to find.

I can confirm that editing `compare_for_fnt` in lib/src/rom/file.rs like this is enough to get rid of all the FNT diffs in 999 (no idea about other games):
```diff
    fn compare_for_fnt(a: &str, a_dir: bool, b: &str, b_dir: bool) -> Ordering {
        let files_first = a_dir.cmp(&b_dir);
        if files_first.is_ne() {
            return files_first;
        }

-        let a_lower = a.to_lowercase();
-        let b_lower = b.to_lowercase();
-        let (a_bytes, _, _) = SHIFT_JIS.encode(&a_lower);
-        let (b_bytes, _, _) = SHIFT_JIS.encode(&b_lower);
+        let (mut a_bytes, _, _) = SHIFT_JIS.encode(&a);
+        let (mut b_bytes, _, _) = SHIFT_JIS.encode(&b);
+        
+        // This is strictly less correct, but matches the ordering of files in 999
+        // Force to lowercase byte-by-byte
+        let mut a_vec = a_bytes.to_mut();
+        let mut b_vec = b_bytes.to_mut();
+        a_vec.make_ascii_lowercase();
+        b_vec.make_ascii_lowercase();

        // Lexicographic, case-insensitive Shift-JIS order
-        a_bytes.cmp(&b_bytes)
+        a_vec.cmp(&b_vec)
    }
```

Regarding the final reported size of the ROM, there appear to be differences in the order that files got inserted, and that screwed up the file allocation table. In the first diff outside of that table that I found, the original ROM had `bg/minigame/a42_morse/resource/main/bg_main.dat`, but the ROM generated by dsrom had the data for `bg/minigame/a42_morse/resource/bg_morse.dat` there instead. This is what path_order.txt looks like there:
```
/bg/minigame/a42_morse/resource/bg_morse.dat
/bg/minigame/a42_morse/resource/subscreen.dat
/bg/minigame/a42_morse/resource/button.dat
/bg/minigame/a42_morse/resource/button_push.dat
/bg/minigame/a42_morse/resource/caret.dat
/bg/minigame/a42_morse/resource/icon_changelist_a.dat
/bg/minigame/a42_morse/resource/icon_changelist_j.dat
/bg/minigame/a42_morse/resource/icon_changelist_s.dat
/bg/minigame/a42_morse/resource/icon_quit.dat
/bg/minigame/a42_morse/resource/icon_reset.dat
/bg/minigame/a42_morse/resource/ton.dat
/bg/minigame/a42_morse/resource/tu.dat
/bg/minigame/a42_morse
```
and that... seems strange to me. It doesn't mention `.../resource/main` at all even though the rest of the files in `.../resource` are mentioned.

**Comment by AetiasHax** [comment]
How strange. The intended behavior is to recurse every path in `path_order.txt` from top to bottom. Usually the path order is pretty short, as the ROM file order is normally sorted lexicographically with the exception of certain directories being recursed before others. For reference, the path order for *The Legend of Zelda: Phantom Hourglass (Europe)* is 26 lines long and was simplified from 3005 files.

I would argue that the simplified path order you posted is correct, but dsrom fails to recurse it correctly somehow... Something about how `/bg/minigame/a42_morse/resource` is split into two parts might be the cause of this bug.

**Comment by PhoenixBound** [comment]
After some more testing... it looks like the issue is some files getting visited twice in this scenario:

![image](https://github.com/user-attachments/assets/b61be7dc-a0c1-46f7-9091-72985e63d7d1)

Now it makes sense why I was seeing the data for bg_morse.dat of all things when it shows up earlier in the path order... The `traverse_files`/`traverse_nonvisited_files` function only keeps track of IDs for directories, not individual files, so it can visit the same file twice.

Tweaking that in a quick-and-dirty way is enough to make 999/BSKJ rebuild properly! (Outside of some checksums being wrong if you generate a decrypted ROM, of course.)
```diff
    fn traverse_nonvisited_files<Cb>(&self, visited: &mut HashSet<u16>, callback: &mut Cb, subdir: &Dir, path: &Path)
    where
        Cb: FnMut(&File, &Path) -> (),
    {
        if visited.contains(&subdir.id) {
            return;
        }
         for child in &subdir.children {
+            if visited.contains(child) {
+                continue;
+            }
+
             if Self::is_dir(*child) {
                 let path = path.join(self.name(*child));
                 self.traverse_nonvisited_files(visited, callback, self.dir(*child), &path);
             } else {
                 callback(self.file(*child), path);
+                if !visited.insert(*child) {
+                    panic!("File {} visited more than once", child);
+                }
             }
         }
-        visited.insert(subdir.id);
+        if !visited.insert(subdir.id) {
+            panic!("Subdirectory {} visited more than once", subdir.id);
+        }
     }

    /// Traverses the [`FileSystem`] and calls `callback` for each file found. The directories will be prioritized according to
    /// the `path_order`.
    pub fn traverse_files<I, Cb>(&self, path_order: I, mut callback: Cb)
    where
        I: IntoIterator<Item = &'a str>,
        Cb: FnMut(&File, &Path) -> (),
    {
        let mut visited = HashSet::<u16>::new();

        for path in path_order {
            let path = path.strip_prefix("/").unwrap_or(path);
            let path_buf = &PathBuf::from_str(path).unwrap();
            let subdir = if path.trim() == "" {
                self.dir(ROOT_DIR_ID)
            } else {
                 let Some(child) = self.find_path(path) else { continue };
+                if visited.contains(&child) {
+                    continue;
+                }
+
                 if Self::is_dir(child) {
                     self.dir(child)
                 } else {
                     let file = self.file(child);
                     callback(file, path_buf);
+                    if !visited.insert(file.id) {
+                        panic!("File {} visited more than once", file.id);
+                    }
                     continue;
                 }
             };
```

**Comment by AetiasHax** [comment]
Thanks for the fixes! It's included in release v0.4.1.
